pkg/doc/jwt/jwt.go

```diff
@@ -0,0 +1,448 @@
+	AlgorithmNone = "none"
+
+	// SignatureEdDSA defines EdDSA alg
+	SignatureEdDSA = "EdDSA"
+
+	// SignatureRS256 defines RS256 alg
+	SignatureRS256 = "RS256"
+)
+
+const (
+	jwtPartsNumber   = 3
+	jwtHeaderPart    = 0
+	jwtPayloadPart   = 1
+	jwtSignaturePart = 2
+)
+
+// JSONWebToken defines JWT basic operations with JWT headers, payload and signature.
+type JSONWebToken interface {
+
+	// Claims fills input c with claims. It could be structure (e.g. Claims or its extension) or map[string]interface{}.
+	Claims(c interface{}) error
+
+	// VerificationData build verification data which can be used for signing or verification.
+	VerificationData() ([]byte, error)
+
+	// Header makes look up of particular header.
+	Header(name string) (interface{}, bool)
+
+	// StringHeader makes look up of particular header with string value.
+	StringHeader(name string) string
+
+	// Signature returns a signature if defined.
+	Signature() []byte
+
+	// CompactSerialize makes compact serialization of token.
+	CompactSerialize(detached bool) (string, error)
+}
+
+type jsonWebToken struct {
+	headers map[string]interface{}
+
+	payload map[string]interface{}
+
+	signature []byte
+}
+
+// CompactSerialize implements compact serialization of token.
+func (j *jsonWebToken) CompactSerialize(detached bool) (string, error) {
+	headers, err := toBase64(j.headers)
+	if err != nil {
+		return "", fmt.Errorf("base64 headers: %w", err)
+	}
+
+	payload := ""
+	if !detached {
+		payload, err = toBase64(j.payload)
+		if err != nil {
+			return "", fmt.Errorf("base64 payload: %w", err)
+		}
+	}
+
+	signature := base64.RawURLEncoding.EncodeToString(j.signature)
+
+	return fmt.Sprintf("%s.%s.%s", headers, payload, signature), nil
+}
+
+// Claims fills input c with JWT claims taken from payload of a token.
+func (j *jsonWebToken) Claims(c interface{}) error {
+	pBytes, err := json.Marshal(j.payload)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(pBytes, c)
+}
+
+// VerificationData build verification data which can be used for signing or verification.
+func (j *jsonWebToken) VerificationData() ([]byte, error) {
+	return jwsVerificationData(j.headers, j.payload)
+}
+
+// Header makes look up of particular header.
+func (j *jsonWebToken) Header(name string) (interface{}, bool) {
+	v, ok := j.headers[name]
+
+	return v, ok
+}
+
+// StringHeader makes look up of particular header with string value.
+func (j *jsonWebToken) StringHeader(name string) string {
+	if headerValue, ok := j.headers[name]; ok {
+		if headerStrValue, ok := headerValue.(string); ok {
+			return headerStrValue
+		}
+	}
+
+	return ""
+}
+
+// Signature returns a copy of JWT signature if defined.
+func (j *jsonWebToken) Signature() []byte {
+	if j.signature == nil {
+		return nil
+	}
+
+	return append(j.signature[:0:0], j.signature...)
+}
+
+// Signer defines JWT Signer interface. It makes signing of data and provides custom JWT headers relevant to the signer.
```

Comment:
we also need pluggable verification so the crypto primitives can be pluggable.

Reply:
`Verifier` interface is defined at separate file `verifier.go`. There is one basic implementation of it created by `NewVerifier()`. It now supports EdDSA and RS256. And currently, there is not way to plug in an extra concrete algorithm verifier. @troyronda  do you want to make it pluggable at this point?

Reply:
we will need to make it pluggable, yes.
(and the default/basic implementations can be overridden.)

Reply:
https://github.com/hyperledger/aries-framework-go/issues/1267